perceval/client.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2015-2017 Bitergia
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
+#
+# Authors:
+#     Valerio Cosentino <valcos@bitergia.com>
+#
+
+import logging
+import time
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+
+from .errors import RateLimitError
+
+logger = logging.getLogger(__name__)
+
+
+class HttpClient:
+    """Abstract class for HTTP clients.
+
+    Base class to query data sources.
+
+    Derivated classes have to implement `call`, `params`, `headers`
+    `pagination`, `has_retry` and `has_sleep` methods. Otherwise, `NotImplementedError`
+    exception will be raised.
+
+    To track which version of the client was used during the fetching
+    process, this class provides a `version` attribute that each client
+    may override.
+
+    :param has_rest: rest between calls
+    :param has_retry: retry calls in case of data source failures
+    :param has_pagination: client returns paginated items from the data source
+    """
+    version = '0.1'
+
+    MIN_RATE_LIMIT = 10
+    MAX_RATE_LIMIT = 500
+
+    DEFAULT_SLEEP_TIME = 1
+    MAX_RETRIES = 5
+
+    MAX_RETRIES_ON_CONNECT = 5
+    MAX_RETRIES_ON_READ = 5
+    MAX_RETRIES_ON_REDIRECT = 5
+    MAX_RETRIES_ON_STATUS = 5
+
+    RATE_LIMIT_HEADER = "X-RateLimit-Remaining"
+    RATE_LIMIT_RESET_HEADER = "X-RateLimit-Reset"
+
+    DEFAULT_RETRY_AFTER_STATUS_CODES = [503, 413, 429]
+
+    GET = "GET"
+    POST = "POST"
+
+    def __init__(self, base_url, sleep_for_rate=False, min_rate_to_sleep=MIN_RATE_LIMIT,
+                 max_retries=MAX_RETRIES, default_sleep_time=DEFAULT_SLEEP_TIME):
+        self.base_url = base_url
+        self.rate_limit = None
+        self.rate_limit_reset_ts = None
+        self.session = None
+
+        self.sleep_for_rate = sleep_for_rate
+
+        if min_rate_to_sleep > self.MAX_RATE_LIMIT:
```

Comment:
I'm still not sure what to do with rate limit handlers. Not sure if we should move this behaviour to a specific class and use it as a decorator or to have a mixin. For instance:
```
class GitHubClient(RateLimitHandler, HttpClient):
    [...]
```

Reply:
We can discuss about it. In your example, the responsabilities of the the HttpClient would be just about connection issues?

Reply:
That's the idea. To create a session, to define which parameters are needed and fetch the items while connection or defined HTTP errors are handled to retry.

When at first I thought of handling rate limit in a generic client, it looked a good idea to me. But now I see it on the code, I feel it a bit weird. So, having a specific class which handles only those cases it's a better idea to me.

What do you think?

Reply:
Let's try with a specific class

Reply:
I added the class **RateLimitHandler** in the client.py. 
The github client remains a subclass of HttpClient -> **GitHubClient(HttpClient)**, while the HttpClient can accept a rate-limit-handler as parameter, which allows to deal with rate limit issues.

This is a bit different from your proposal (**GitHubClient(RateLimitHandler, HttpClient)**), but I'd would say that the generic handler should be coupled with the generic client. In case a given client wants to use the handler, it will have to set up the handler and pass it to the generic client, like this:

```python
rate_limit_handler = RateLimitHandler(sleep_for_rate, min_rate_to_sleep)
super().__init__(self.api_url, default_sleep_time=default_sleep_time, max_retries=max_retries,
                 rate_limit_handler=rate_limit_handler)
```

your thoughts?